pkg/client/access_control_records_object_set.go

```diff
@@ -3,90 +3,95 @@
 package client
 
 import (
-	"reflect"
 	"fmt"
 	"github.com/hpe-storage/common-host-libs/jsonutil"
-	"github.com/hpe-storage/nimble-golang-sdk/pkg/client/v1/model"
-	"github.com/hpe-storage/nimble-golang-sdk/pkg/util"
+	"github.com/hpe-storage/nimble-golang-sdk/pkg/client/v1/nimbleos"
+	"github.com/hpe-storage/nimble-golang-sdk/pkg/param"
+	"reflect"
 )
 
-
 // Manage access control records for volumes.
 const (
-    accessControlRecordPath = "access_control_records"
+	accessControlRecordPath = "access_control_records"
 )
 
 // AccessControlRecordObjectSet
 type AccessControlRecordObjectSet struct {
-    Client *GroupMgmtClient
+	Client *GroupMgmtClient
 }
 
 // CreateObject creates a new AccessControlRecord object
-func (objectSet *AccessControlRecordObjectSet) CreateObject(payload *model.AccessControlRecord) (*model.AccessControlRecord, error) {
-	accessControlRecordObjectSetResp, err := objectSet.Client.Post(accessControlRecordPath, payload)
-	if err !=nil {
-		return nil,err
+func (objectSet *AccessControlRecordObjectSet) CreateObject(payload *nimbleos.AccessControlRecord) (*nimbleos.AccessControlRecord, error) {
+	encodedPayload, err := nimbleos.EncodeAccessControlRecord(payload)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := objectSet.Client.Post(accessControlRecordPath, encodedPayload)
```

Comment:
Subsequent code trusts that the "resp" value is an accessControlRecord structure.  Why?  Because the encodedPayload we pass in is an accessControlRecord structure, and Post returns the same type it is given.  (Note: This needs to be a documented behavior.)  By the same token, why wouldn't we trust that "resp" is not nil?  I think you are saying it is not practical to verify the response type but it is practical to verify a non-nil response.  I'd like to understand the distinction.

The Post function is able to return the correct type because it simply returns a filled in copy of the structure it is given.  (Or is it actually the same structure?)  I wonder if this could be problematic in some cases.  I suspect the JSON unmarshal logic leaves the existing structure contents in place and simply overwrites attributes as it decodes them.  Is that correct?  If so, it might be mishandling write-only attributes.  Consider a "set password" API.  The password value would be included in the request but not in the response.  If the response starts out with a copy of the request, it would incorrectly contain the password value.

Perhaps all of these items could be addressed by adding a responseHolder parameter to the Post function, which the caller must provide as a structure of the correct type without any values set.  What do you think of that idea?

It isn't clear to me how asynchronous responses (202 Accepted) will work, so maybe my suggestion won't be great for that case.  List actions could seemingly clone the responseHolder as necessary, though it isn't clear to me whether List is fully implemented yet.

Reply:
> Subsequent code trusts that the "resp" value is an accessControlRecord structure. Why? Because the encodedPayload we pass in is an accessControlRecord structure, and Post returns the same type it is given. 

POST doesn't return the same type as it is given, it returns the interface{} type. Take a look at group_mgmt_client.go 

>By the same token, why wouldn't we trust that "resp" is not nil?
This is additional check to ensure it doesn't lead to application crash. 

>It isn't clear to me how asynchronous responses (202 Accepted) will work,
It is still not implemented. SDK is work in progress and there are many things in the pipeline to be implemented. 

>Perhaps all of these items could be addressed by adding a responseHolder parameter to the Post function, which the caller must provide as a structure of the correct type without any values set. What do you think of that idea?

I feel this is unnecessary as we don't need to pass the struct type .

Reply:
> POST doesn't return the same type as it is given, it returns the interface{} type. Take a look at group_mgmt_client.go

I see that Post is given an interface{} and it returns an interface{}.  It doesn't document what they are or how they are related, but it looks to me like the returned type is the same as the given type.  How do they differ?

> > By the same token, why wouldn't we trust that "resp" is not nil?

> This is additional check to ensure it doesn't lead to application crash.

A failed type assertion also causes a panic, yes?

> I feel this is unnecessary as we don't need to pass the struct type .

I don't understand.  Is my suspicion correct that the returned structure might contain leftover attributes from the request?

Reply:
You should check the POST implementation not simply the return type interface{}. POST is invoked by different objectset which has different types. Internally it uses DataWrapper to extract the POST response . Interface{} is the best way to handle in Go lang. Why would you want to invoke POST with a separate parameter as data type  along with the actual field?

Reply:
I can see that I am not communicating well.  I apologize for that.

Notice how the Get function has an "intf" parameter even though it does not require an input structure.  The intf is essentially used as an empty structure of the correct type, into which the response is unmarshalled (using a DataWrapper, as you mentioned).  I'm suggesting that perhaps Post could have a parameter for the "full" request structure (as it does now) and a separate parameter for the "empty" response structure.  It would address my theorized concern about leftover request attributes being incorrectly included in the response structure.

Furthermore, I am wondering if it would make sense for the response structure to be modified in place, with the corresponding function return value removed.  The client could do something like this:

  requestPayload, err := nimbleos.EncodeAccessControlRecord(payload)
  ... (check err)
  responsePayload := &accessControlRecord{}
  err := objectSet.Client.Post(accessControlRecordPath, requestPayload, responsePayload)
  ... (check err)
  // Note: The DecodeAccessControlRecord parameter could be *accessControlRecord instead of interface{}.
  return nimbleos.DecodeAccessControlRecord(responsePayload)

There would be no way for responsePayload to be nil or to be of an incorrect type.  We wouldn't need logic to check for either, and we could remove the type assertion that could theoretically cause a panic.

Reply:
>Notice how the Get function has an "intf" parameter even though it does not require an input structure. The intf is essentially used as an empty structure of the correct type, into which the response is unmarshalled (using a DataWrapper, as you mentioned). I'm suggesting that perhaps Post could have a parameter for the "full" request structure (as it does now) and a separate parameter for the "empty" response structure. It would address my theorized concern about leftover request attributes being incorrectly included in the response structure.

Now I see your point but that will only happen if API response is nil for the concerned attribute and HTTP return a SUCCESS since the same variable payload is used for unmarshalling as well. Although, it is a corner case but still valid. 

thanks for providing more details. I will fix it.